crates/router/src/core/webhooks/utils.rs

```diff
@@ -154,3 +158,98 @@ pub(crate) fn get_idempotent_event_id(
 pub(crate) fn generate_event_id() -> String {
     common_utils::generate_time_ordered_id("evt")
 }
+
+pub async fn perform_redis_lock<A>(
+    state: &A,
+    unique_locking_key: String,
+    merchant_id: common_utils::id_type::MerchantId,
+) -> RouterResult<bool>
+where
+    A: SessionStateInfo,
+{
+    let redis_conn = state
+        .store()
+        .get_redis_conn()
+        .change_context(errors::ApiErrorResponse::InternalServerError)?;
+
+    let redis_locking_key = format!(
+        "{}_{}_{}",
+        WEBHOOK_LOCK_PREFIX,
+        merchant_id.get_string_repr(),
+        unique_locking_key
+    );
+    let redis_lock_expiry_seconds = state.conf().lock_settings.redis_lock_expiry_seconds;
+
+    let redis_lock_result = redis_conn
+        .set_key_if_not_exists_with_expiry(
+            &redis_locking_key.as_str().into(),
+            state.get_request_id(),
+            Some(i64::from(redis_lock_expiry_seconds)),
+        )
+        .await;
+
+    match redis_lock_result {
+        Ok(redis::SetnxReply::KeySet) => {
+            logger::info!("Lock acquired for {:?}", redis_locking_key);
+            Ok(true)
+        }
+        Ok(redis::SetnxReply::KeyNotSet) => {
+            logger::info!("Lock already held for {:?}", redis_locking_key);
+            Ok(false)
+        }
+        Err(err) => Err(err.change_context(errors::ApiErrorResponse::InternalServerError)),
+    }
+}
+
+pub const WEBHOOK_LOCK_PREFIX: &str = "WEBHOOK_LOCK";
+
+pub async fn free_redis_lock<A>(
+    state: &A,
+    unique_locking_key: String,
+    merchant_id: common_utils::id_type::MerchantId,
+) -> RouterResult<()>
+where
+    A: SessionStateInfo,
+{
+    let redis_conn = state
+        .store()
+        .get_redis_conn()
+        .change_context(errors::ApiErrorResponse::InternalServerError)?;
+
+    let redis_locking_key = format!(
+        "{}_{}_{}",
+        WEBHOOK_LOCK_PREFIX,
+        merchant_id.get_string_repr(),
+        unique_locking_key
+    );
+    match redis_conn
+        .get_key::<Option<String>>(&redis_locking_key.as_str().into())
+        .await
+    {
+        Ok(val) => {
+            if val == state.get_request_id() {
+                match redis_conn
+                    .delete_key(&redis_locking_key.as_str().into())
+                    .await
+                {
+                    Ok(redis::types::DelReply::KeyDeleted) => {
+                        logger::info!("Lock freed {:?}", redis_locking_key);
```

Comment:
Can use `Display` impl here.

```suggestion
                        logger::info!("Lock freed {redis_locking_key}");
```